pkg/controller/clone-controller.go

```diff
@@ -203,108 +230,382 @@ func (cc *CloneController) waitTargetPodRunningOrSucceeded(pvc *v1.PersistentVol
+	if !exists {
+		return nil, errors.Errorf("bad CloneRequest Annotation")
+	}
+
+	ownerKey, err := cache.MetaNamespaceKeyFunc(pvc)
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "error getting cache key")
 	}
 
-	cc.podExpectations.DeleteExpectations(key)
+	serverCACertBytes, err := keys.GetKeyPairAndCertBytes(r.K8sClient, util.GetNamespace(), uploadServerCASecret)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting uploadserver server CA cert")
+	}
 
-	return nil
-}
+	if serverCACertBytes == nil {
+		return nil, errors.Errorf("secret %s does not exist", uploadServerCASecret)
+	}
 
-func (cc *CloneController) validateSourceAndTarget(targetPvc *v1.PersistentVolumeClaim) error {
-	sourcePvc, err := getCloneRequestSourcePVC(targetPvc, cc.Controller.pvcLister)
+	clientCAKeyPair, err := keys.GetKeyPairAndCert(r.K8sClient, util.GetNamespace(), uploadServerClientCASecret)
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "error getting uploadserver client CA cert")
 	}
 
-	if err = validateCloneToken(cc.tokenValidator, sourcePvc, targetPvc); err != nil {
-		return err
+	if clientCAKeyPair == nil {
+		return nil, errors.Errorf("secret %s does not exist", uploadServerClientCASecret)
 	}
 
-	return ValidateCanCloneSourceAndTargetSpec(&sourcePvc.Spec, &targetPvc.Spec)
+	clientKeyBytes, clientCertBytes, err := createClientKeyAndCertFunc(&clientCAKeyPair.KeyPair, clientName, []string{})
+	if err != nil {
+		return nil, err
+	}
+
+	podResourceRequirements, err := GetDefaultPodResourceRequirements(r.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	pod := MakeCloneSourcePodSpec(image, pullPolicy, sourcePvcName, sourcePvcNamespace, ownerKey, clientKeyBytes, clientCertBytes, serverCACertBytes.Cert, pvc, podResourceRequirements)
+
+	if err := r.Client.Create(context.TODO(), pod); err != nil {
+		return nil, errors.Wrap(err, "source pod API create errored")
+	}
+
+	log.V(1).Info("cloning source pod (image) created\n", "pod.Namespace", pod.Namespace, "pod.Name", pod.Name, "image", image)
+
+	return pod, nil
+}
+
+func getCloneSourceName(sourcePvcName string) string {
+	return fmt.Sprintf("%s-%s-", common.ClonerSourcePodName, sourcePvcName)
+}
+
+// MakeCloneSourcePodSpec creates and returns the clone source pod spec based on the target pvc.
+func MakeCloneSourcePodSpec(image, pullPolicy, sourcePvcName, sourcePvcNamespace, ownerRefAnno string,
+	clientKey, clientCert, serverCACert []byte, pvc *corev1.PersistentVolumeClaim, resourceRequirements *corev1.ResourceRequirements) *corev1.Pod {
+
+	var ownerID string
+	podName := getCloneSourceName(sourcePvcName)
+	id := string(pvc.GetUID())
+	url := GetUploadServerURL(pvc.Namespace, pvc.Name)
+	pvcOwner := metav1.GetControllerOf(pvc)
+	if pvcOwner != nil && pvcOwner.Kind == "DataVolume" {
+		ownerID = string(pvcOwner.UID)
+	}
+
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: podName,
```

Comment:
I believe using GenerateName without the protection of something like expectations can get you multiple pods.   But I'm not exactly sure how.  Maybe it's safe.  Otherwise, some thought would have to go into how to get a deterministic name here.  Not as simple as import/upload case because there could theoretically be multiple concurrent clones with the same source.  Which is something we should probably be testing if we aren't.

Reply:
We do have a functional test for that: https://github.com/kubevirt/containerized-data-importer/blob/master/tests/cloner_test.go#L395
When we look up the source pod, we have a label selector that uses the UID of the TARGET pvc, and thus it doesn't matter if we have multiple source pods, the lookup will find only the one we want based on the label selector.

Reply:
I understand how that works but I think it is possible to create a pod then list pods and not get the pod you just created.  Especially with controller runtime client which is cached.

Reply:
I get what you are saying, it is possible to list (for instance in a different reconcile loop) before the create completed, and thus you won't find the pod, and create another one. The only way I can think of to get a unique deterministic name is to use the target pvc name and target pvc namesapce. If we just do source-<tgt_name>-<tgs_ns> its likely we will run into the name length limit of the pod pretty quickly. So I would prefer to do something else, maybe a hash of those two?

Reply:
Decided to do a hash, I updated the PR.